ludwig/models/trainer.py

```diff
@@ -738,93 +738,7 @@ def train(self, training_set, validation_set=None, test_set=None, save_path="mod
-                    else:
-                        current_learning_rate = learning_rate_warmup(
-                            current_learning_rate,
-                            progress_tracker.epoch,
-                            self.learning_rate_warmup_epochs,
-                            batcher.step,
-                            batcher.steps_per_epoch,
-                        )
-                    self.set_learning_rate(current_learning_rate)
-
-                    # obtain batch
-                    batch = batcher.next_batch()
-
-                    # Move tensors to cuda here.
-                    inputs = {
-                        i_feat.feature_name: torch.from_numpy(batch[i_feat.proc_column]).to(self.device)
-                        for i_feat in self.model.input_features.values()
-                    }
-                    targets = {
-                        o_feat.feature_name: torch.from_numpy(batch[o_feat.proc_column]).to(self.device)
-                        for o_feat in self.model.output_features.values()
-                    }
-
-                    # Reintroduce for tensorboard graph
-                    # if first_batch and self.is_coordinator() and not skip_save_log:
-                    #    tf.summary.trace_on(graph=True, profiler=True)
-
-                    loss, all_losses = self.train_step(
-                        inputs,
-                        targets,
-                    )
-
-                    # Reintroduce for tensorboard graph
-                    # if first_batch and self.is_coordinator() and not skip_save_log:
-                    #     with train_summary_writer.as_default():
-                    #         tf.summary.trace_export(
-                    #             name="Model",
-                    #             step=0,
-                    #             profiler_outdir=tensorboard_log_dir
-                    #         )
-
-                    if self.is_coordinator() and not self.skip_save_log:
-                        self.write_step_summary(
-                            train_summary_writer=train_summary_writer,
-                            combined_loss=loss,
-                            all_losses=all_losses,
-                            step=progress_tracker.steps,
-                            learning_rate=current_learning_rate,
-                        )
-
-                    progress_tracker.steps += 1
-                    if self.is_coordinator():
-                        progress_bar.update(1)
-                        logger.debug(
-                            f"training: completed batch {progress_bar.n} "
-                            f"memory used: "
-                            f"{psutil.Process(os.getpid()).memory_info()[0] / 1e6:0.2f}MB"
-                        )
-
-                    self.callback(lambda c: c.on_batch_end(self, progress_tracker, save_path))
+                self._train_loop(batcher, progress_tracker, save_path, train_summary_writer, progress_bar)
```

Comment:
At some point, we may want to re-think the epoch-based training-eval loops as we begin to work with large datasets (O(10) or O(100) million examples).

A nice setup, in my experience, was:
1. The trainer outputs checkpoints every few thousand steps/batches to a checkpoint directory.
2. A separate continuous eval process (binary?) waits for new files in the checkpoint directory.
3. When a new checkpoint gets written, the eval process runs eval on the eval set. This can be configured to either:
    - predict on a fixed number of examples (or the entire eval set), output metrics, wait for / grab the latest checkpoint, repeat
    - predict as many batches as possible until the next checkpoint gets written, output metrics, repeat

    The latter is nice for having eval metrics available for every checkpoint, but the former is better for comparing across modeling runs.

4. In either mode, at the end of training, metrics are computed over the entire eval set using the final checkpoint.

@w4nderlust @ShreyaR FYI

Reply:
Yes makes sense, let's do a brainstorming session on this after v0.5